src/net.cpp

```diff
@@ -1573,26 +1579,55 @@ void CConnman::ThreadDNSAddressSeed()
     if (gArgs.GetBoolArg("-forcednsseed", DEFAULT_FORCEDNSSEED)) {
         // When -forcednsseed is provided, query all.
         seeds_right_now = seeds.size();
+    } else if (addrman.size() == 0) {
+        // If we have no known peers, query all.
+        seeds_right_now = seeds.size();
     }
 
+    // goal: only query DNS seed if address need is acute
+    // * If we have a reasonable number of peers in addrman, spend
+    //   some time trying them first. This improves user privacy by
+    //   creating fewer identifying DNS requests, reduces trust by
+    //   giving seeds less influence on the network topology, and
+    //   reduces traffic to the seeds.
+    // * When querying DNS seeds query a few at once, this ensures
+    //   that we don't give DNS seeds the ability to eclipse nodes
+    //   that query them.
+    // * If we continue having problems, eventually query all the
+    //   DNS seeds, and if that fails too, also try the fixed seeds.
+    //   (done in ThreadOpenConnections)
+    const std::chrono::seconds seeds_wait_time = (addrman.size() >= DNSSEEDS_DELAY_PEER_THRESHOLD ? DNSSEEDS_DELAY_MANY_PEERS : DNSSEEDS_DELAY_FEW_PEERS);
+
     for (const std::string& seed : seeds) {
-        // goal: only query DNS seed if address need is acute
-        // Avoiding DNS seeds when we don't need them improves user privacy by
-        // creating fewer identifying DNS requests, reduces trust by giving seeds
-        // less influence on the network topology, and reduces traffic to the seeds.
         if (addrman.size() > 0 && seeds_right_now == 0) {
-            if (!interruptNet.sleep_for(std::chrono::seconds(11))) return;
-
-            LOCK(cs_vNodes);
-            int nRelevant = 0;
-            for (const CNode* pnode : vNodes) {
-                nRelevant += pnode->fSuccessfullyConnected && !pnode->fFeeler && !pnode->fOneShot && !pnode->m_manual_connection && !pnode->fInbound;
-            }
-            if (nRelevant >= 2) {
-                LogPrintf("P2P peers available. Skipped DNS seeding.\n");
-                return;
+            LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
+            std::chrono::seconds to_wait = seeds_wait_time;
+            while (to_wait.count() > 0) {
+                std::chrono::seconds w{25}; // check for early abort every 25 seconds
```

Comment:
As 25 seconds is a timeout for the _early_ exit from dnsseed thread, we could do not introduce a new heuristic constant at all:
```diff
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -41,9 +41,9 @@
 static_assert(MINIUPNPC_API_VERSION >= 10, "miniUPnPc API version >= 10 assumed");
 #endif
 
-#include <unordered_map>
-
+#include <chrono>
 #include <math.h>
+#include <unordered_map>
 
 // Dump addresses to peers.dat every 15 minutes (900s)
 static constexpr int DUMP_PEERS_INTERVAL = 15 * 60;
@@ -1601,12 +1601,12 @@ void CConnman::ThreadDNSAddressSeed()
     for (const std::string& seed : seeds) {
         if (addrman.size() > 0 && seeds_right_now == 0) {
             LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
-            std::chrono::seconds to_wait = seeds_wait_time;
+            std::chrono::milliseconds to_wait{seeds_wait_time};
             while (to_wait.count() > 0) {
-                std::chrono::seconds w{25}; // check for early abort every 25 seconds
-                if (w > to_wait) w = to_wait;
-                if (!interruptNet.sleep_for(w)) return;
-                to_wait -= w;
+                std::chrono::milliseconds recheck_period{nConnectTimeout};
+                if (recheck_period > to_wait) recheck_period = to_wait;
+                if (!interruptNet.sleep_for(recheck_period)) return;
+                to_wait -= recheck_period;
 
                 int nRelevant = 0;
                 {
```

Reply:
It'd make sense for `DNSSEEDS_DELAY_FEW_PEERS` to be calculated from that rather than a constant, but the early exit ought to be based on how long it takes for a connection to _succeed_ rather than the timeout time

Reply:
On the other hand, can easily reuse `DNSSEEDS_DELAY_FEW_PEERS` instead of a new constant, so will just do that.